packages/eslint-plugin-react-hooks/src/RulesOfHooks.js

```diff
@@ -110,47 +110,29 @@ export default {
          * Segments 1 and 2 have one path to the beginning of `MyComponent` and
          * segment 3 has two paths to the beginning of `MyComponent` since we
          * could have either taken the path of segment 1 or segment 2.
-         *
-         * Populates `cyclic` with cyclic segments.
          */
 
-        function countPathsFromStart(segment) {
-          const {cache} = countPathsFromStart;
```

Comment:
The reason I use a cache here is that in complex components, time complexity can really start to spike up. Consider:

```js
function MyComponent() {
  if (a) {} else {}
  if (b) {} else {}
  if (c) {} else {}
  useHook();
}
```

Here we have 2<sup>3</sup> = 8 paths from `useHook()` to the start of `MyComponent`. Representing the following combinations of `a`, `b`, and `c`.

|a|b|c|
|-|-|-|
|true|true|true|
|false|true|true|
|true|false|true|
|false|true|false|
|true|false|false|
|false|true|false|
|false|false|true|
|false|false|false|

So we have a 2<sup>n</sup> exponential relationship, fun. Now remember that every `&&` and `||` (in the future perhaps also [`??`](https://github.com/tc39/proposal-nullish-coalescing)) introduces a condition since `false && expensive()` will not execute `expensive()`.

Let’s say we have a complex component that has 5 conditions placed in the component before 6 hooks all in the same segment. Without a cache we have to call `countPathsFromStart()` on 2<sup>5</sup> paths 6 times. With a cache we only need to call `countPathsFromStart()` on 2 × 5 segments because we cache the value for every segment so we only need to visit each segment once.

In big-O notation where “n” is the number of conditions and “h” is the number of hooks we have O(2n) time complexity with a cache and O(2<sup>n</sup> × h) without a cache.

To see what I mean in practice try adding the following component to the test suite. On this branch I waited about 10s and the test suite hadn’t finished. When I switched back to master the test suite finished in about 3s.

```js
function MyComponent() {
  // 40 conditions
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}
  if (c) {} else {}

  // 10 hooks
  useHook();
  useHook();
  useHook();
  useHook();
  useHook();
  useHook();
  useHook();
  useHook();
  useHook();
  useHook();
}
```

Reply:
Hi! Thanks for the chime in, explaining the original rationale behind this piece of code, that was really enlightening.

Unfortunately, I'm not sure there's a way to work around this performance regression. Using the cache as is proved itself faulty, as issue proved.
I could just remove the item from the cache once we finished visiting it, but the complexity of the overall algorithm would be the same.
I guess the complexity of the problem of finding the number of paths between two nodes on a graph can't really be reduced here.

I'll think about how we could improve it for the case of multiple hooks, so the complexity is not so high.

Reply:
I added some components below with a possible solution and the rationale behind that solution. TL;DR when `countPathsFromStart()` breaks a cycle it gives a temporary result of 0 to avoid looping forever.